GitCommands/Git/GitModule.cs

```diff
@@ -2185,13 +2185,8 @@ public Patch GetSingleDiff(string @from, string to, string fileName, string oldF
 
             //fix refs slashes
             from = from.ToPosixPath();
-            to = to.ToPosixPath();
-            string commitRange = string.Empty;
-            if (!to.IsNullOrEmpty())
-                commitRange = "\"" + to + "\"";
-            if (!from.IsNullOrEmpty())
-                commitRange = string.Join(" ", commitRange, "\"" + from + "\"");
-
+            to = to == null ? "":to.ToPosixPath();
+            string commitRange = RevisionDiffProvider.Get(from, to);
```

Comment:
RevisionDiffProvider.Get() is really an instance function to GitModule. A separate instance would add no value, it holds no status and has no external dependencies.

What is better with (new RevisionDiffProvider ()).Get(from, to)?

Reply:
A SOLID implementation does not hold state and may or may not have external dependencies. A Following SOLID principles all dependencies are interfaces-based (O) and injected via a constructor (D).
This allows type substitution (L) and unit testing - ensuring that the implementation under test is correct and not that all dependencies behave correctly. This is done in integration tests.
Unit and integration tests compliment each other. Though some code paths maybe be easier to only unit tests and some only integration test.

The target state I would like to get to is to have SOLID GitModule - no state, just pure behaviors. With all dependencies injected via the constructor.
It should simplify a lot of code paths through the app.


> What is better with (new RevisionDiffProvider ()).Get(from, to)?

This what I expect to see:

```
    public sealed class GitModule : IGitModule
    {
        // ...
        private readonly IRevisionDiffProvider _revisionDiffProvider = new RevisionDiffProvider();

        public Patch GetSingleDiff(string @from, string to, string fileName, string oldFileName, string extraDiffArguments, Encoding encoding, bool cacheResult)
        {
            // ...

            //fix refs slashes
            from = from.ToPosixPath();
            to = to == null ? "":to.ToPosixPath();
            string commitRange = _revisionDiffProvider.Get(from, to);
            if (AppSettings.UsePatienceDiffAlgorithm)
                extraDiffArguments = string.Concat(extraDiffArguments, " --patience");

            // ...
        }

}

```

Reply:
Adding an instance for _revisionDiffProvider is just making the code hard to read an therefore to maintain, it adds an object that does not hold anything. To follow SOLID for this is overthinking.
The only reason to have RevisionDiffProvider as a separate class instead of a private instance function in GitModule is the testability. Maybe the tests can be done in GitModuleTestHelper instead, but that would decrease the separation.

I prefer  (new RevisionDiffProvider ()).Get(from, to) in this case.

Reply:
@RussKie RevisionDiffProvider uses static methods in RevisionGit class. How to handle those if RevisionDiffProvider is made an instance?

Reply:
Made instance class. Cannot change "const string" in GitRevision as it affects the test cases